lib/spack/llnl/util/filesystem.py

```diff
@@ -2236,30 +2236,27 @@ def link_dest(self):
         if hasattr(self.pkg, "libs") and self.pkg.libs:
             pkg_libs = set(self.pkg.libs.directories)
         else:
-            pkg_libs = set((self.pkg.prefix.lib, self.pkg.prefix.lib64))
+            pkg_libs = set((self.pkg.prefix.lib, self.pkg.prefix.lib64, self.pkg.prefix.bin))
+        return pkg_libs
 
-        return pkg_libs | set([self.pkg.prefix.bin]) | self.internal_links
-
-    @property
-    def internal_links(self):
-        """
-        linking that would need to be established within the package itself. Useful for links
-        against extension modules/build time executables/internal linkage
-        """
-        return self._internal_links
-
-    def add_internal_links(self, *dest):
+    def add_library_dependent(self, *dest):
         """
-        Incorporate additional paths into the rpath (sym)linking scheme.
-
-        Paths provided to this method are linked against by a package's libraries
-        and libraries found at these paths are linked against a package's binaries.
-        (i.e. /site-packages -> /bin and /bin -> /site-packages)
+        Add paths to set of common paths that need to link against other libraries
 
-        Specified paths should be outside of a package's lib, lib64, and bin
+        Specified paths should be outside of a package's common link paths, i.e. the lib, lib64, and bin
         directories.
         """
-        self._internal_links = self._internal_links | set(*dest)
+        self._internal_links = self._internal_links | set(dest)
+
+    @property
+    def internal_link_targets(self):
+        return set(
+            [
+                x
+                for internal_path in self.link_dest
+                for x in find_all_shared_libraries(internal_path, recursive=True)
+            ]
```

Comment:
Can you explain what this list comprehension is doing or expand it out (actually expanding it into multiple statements is probably best)? I don't get the two consecutive `for` statements.

Reply:
The list comprehension is iterating over each directory in `self.link_dest`, searching each of those directories with `find_all_shared_libraries` and then storing each element from the list that call returns. It's the same as 
```
for directory in self.link_dest:
   for lib in find_all_shared_libraries(directory):
      name_of_set.add(lib)
```

Would you still like that expressed more verbosely?

Reply:
Yes: I think most folks will find it more readable (I definitely will).

Reply:
I ended up removing the function entirely. I've added the results of `add_library_dependents` to the `library_dependents` property, so now we have a more consolidated expression of the libraries that need linking vs the rpaths we're linking against. Also this way only requires a single loop in the `establish_link` phase.